src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,822 @@
+ * On eofbit:
+ *      dio::iostream_eof("End Of File has been reached.")
+ * On buffer_failbit:
+ *      dio::iostream_error("Cannot use stream's buffer: buffer is nullptr!")
+ * On string_failbit:
+ *      dio::iostream_error("Cannot put lines into given std::string.")
+ * On posixio_errno:
+ *      dio::posixio_error(errno)
+ *
+ * Usage:
+ *
+ *      // Getting a line from stdin via getline_nothrow()
+ *      while(getline_nothrow(cin, my_string)) {
+ *          my_output.write(my_string, "\n");
+ *      }
+ *
+ *      // Getting a character from stdin via getchar()
+ *      int a_character = getchar();
+ *      if (a_character == -1) return; // Check for EOF or error
+ *      my_output.write((char)a_character, "\n");
+ */
+
+constexpr unsigned IOSTREAM_BUFSIZE = 16384; // Size of the buffers
+
+namespace dio {
+
+/*
+ * Stream state bits/vars:
+ *
+ * 1. "eofbit" indicates that the End Of File has been reached, and there is nothing
+ *    left in the buffer (filein only).
+ * 2. "buffer_failbit" indicates an attempt to use the buffer when the buffer pointer
+ *    was nullptr (e.g. failure to allocate a unique_ptr of streambuf) (fileout/filein).
+ * 3. "string_failbit" indicates a failed operation in std::string related calls
+ *    (e.g. failure in pushing a line from the buffer to the given std::string in
+ *    get_line()) (filein only).
+ * 4. "posixio_errno" indicates that a POSIX I/O function failed, and the POSIX errno
+ *    was set (fileout/filein).
+ *
+ * If all of the above bits/variables are false or 0, the stream status is "good".
+ *
+ * The current_state() function returns the stream's current status based on this table.
+ */
+enum diostates
+{
+    eofbit = 0x01,
+    buffer_failbit = 0x02,
+    string_failbit = 0x04,
+    posixio_errno = 0x08
+};
+
+// Some empty classes for buffer operations implementation.
+// Used for accepting endl and flush in fileout::write() functions.
+class endline { };
+class flushbuf { };
+static constexpr endline endl;   // same as std::endl
+static constexpr flushbuf flush; // same as std::flush
+
+/* Specialise system_error & runtime_error for more meaningful exceptions on state bits/vars */
+// on eofbit
+class iostream_eof : public std::runtime_error
+{
+    public:
+    iostream_eof(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+// on buffer_failbit, string_failbit
+class iostream_error : public std::runtime_error
+{
+    public:
+    iostream_error(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+// on posixio_errno
+class posixio_error : public std::system_error
+{
+    public:
+    posixio_error(const int posix_errno) : system_error(posix_errno, std::system_category())
+    {
+    }
+};
+
+using streambuf = cpbuffer<IOSTREAM_BUFSIZE>;
+
+// fileinout_base is a base for fileout/filein which defines some general things.
+// Public functions are available in both fileout/filein and they does not get defined
+// at those classes.
+class fileinout_base
```

Comment:
By correct convention this would be `file_in_out_base`. But I prefer something like `io_base` or `iostream_base`.

Reply:
Changed to `io_base`